### ISSUE TITLE
kubernetes-client api versions compaibility with runtime api server 

### ISSUE THREAD

**Issue by chakravarthyvp** [fabric8io/kubernetes-client#689]
Hi,

    I'm using 
```
 <groupId>io.fabric8</groupId>
            <artifactId>kubernetes-client</artifactId>
            <version>2.2.0</version>
```
and I have minikube and kubectl installed with versions below:

```
 minikube version
minikube version: v0.16.0
qchavar@elxa1h67k32:~/tools/tmp$ kubectl version
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3", GitCommit:"029c3a408176b55c30846f0faedf56aae5992e9b", GitTreeState:"clean", BuildDate:"2017-02-15T06:40:50Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"1970-01-01T00:00:00Z", GoVersion:"go1.7.1", Compiler:"gc", Platform:"linux/amd64"}

```

Here is my spark job spec., that is generated through the JobBuilder API and dumped as YAML

```
apiVersion: "batch/v2alpha1"
kind: "Job"
metadata:
  annotations: {}
  finalizers: []
  labels: {}
  name: "Test-PBA"
  ownerReferences: []
spec:
  template:
    spec:
      containers:
      - args: []
        command: []
        env: []
        image: "somedomain/aia-spark-rasr:aia-spark-rasr"
        imagePullPolicy: "Always"
        ports: []
        volumeMounts: []
      imagePullSecrets: []
      nodeSelector: {}
      volumes: []
```

This fails with validation when I try to create the job... 
```

kubectl create -f spark-job.yaml
error: error validating "spark-job.yaml": error validating data: field name: is required; if you choose to ignore these errors, turn validation off with --validate=false

```
I changed the api version to batch/v1 as well but it's the same.

Can someone help here with what kubernetes-client api library version to use if that's what is wrong?

The whole of version compatibility with the library against the api server is confusing...

Best Regards
Chakravarthy

**Comment by foxish** [comment]
You're missing 3 things.

* metadata.name: Invalid value: "Test-PBA": must match the regex [a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)* (e.g. 'example.com')
* spec.template.spec.containers[0].name: Required value
* spec.template.spec.restartPolicy: Unsupported value: "Always": supported values: OnFailure, Never

**Comment by chakravarthyvp** [comment]
Hi Foxish,
   I sorted through these issues. However the general issue that I noticed is that there is no version matrix between fabri8 kubernetes-client API and kubernetes API server versions supported. For ex., v1.5.x has moved jobs into batch/v1 however the kubernetes fabric8 version 2.2.x by default uses v1beta1 version of the job group. I have to manually set the API version before talking to the API server...
Essentially there is no version traceability in terms of supported api server version.

Best Regards
Chakravarthy
